cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -1931,14 +1931,29 @@ TEST(Cast, DictTypeToAnotherDict) {
              R"(["a", "b", "a", null])");
 
   auto arr = ArrayFromJSON(dictionary(int32(), int32()), "[1, 1000]");
-  // check unsafe
+  // check unsafe values
   ASSERT_OK_AND_ASSIGN(auto casted,
                        Cast(arr, dictionary(int8(), int8()), CastOptions::Unsafe()));
   ValidateOutput(casted);
-  // check safe
+  // check safe values
   EXPECT_RAISES_WITH_MESSAGE_THAT(
       Invalid, testing::HasSubstr("Integer value 1000 not in range"),
       Cast(arr, dictionary(int8(), int8()), CastOptions::Safe()));
+
+  // check unsafe indices
+  random::RandomArrayGenerator rand(/*seed=*/0);
+  int64_t len = 1000;
+  auto val_arr = rand.ArrayOf(int32(), len, /*null_probability=*/0.01);
+  ASSERT_OK_AND_ASSIGN(auto arr2, DictionaryEncode(val_arr));
+  // check unsafe indices. Cannot validate this array because ValidateOutput throws an
```

Comment:
Sorry - now that I look at the JIRA again, it seems having unsafe casts isn't useful for this case? What do you think? If it produces invalid output (presumably, negative and out of bounds indices?), then it seems kind of pointless.

Reply:
I was also thinking about this. OTOH I was thinking whether the ValidataOutput should check the validity of dictionary array in DictionaryType. :thinking:

Reply:
It looks like it does - are you saying it shouldn't?

Reply:
I dont think ValidateOutput checks the validity of `ArrayData::dictionary` field. That's why ValidateOutput passes for unsafe valies AFAIU. :thin

Reply:
It looks like it does: https://github.com/apache/arrow/blob/dbeed527dff79f3fa4576722a821fb7af3abd878/cpp/src/arrow/array/validate.cc#L565-L572

Reply:
I think there's no way a primitive array to be validated once the casting is complete (unsafely in this instance).

Reply:
We're casting from dictionary to dictionary here though. 

Anyways, the point stands: this is an unsafe cast that generates an invalid array, so is this case worth supporting?

Reply:
I think I agree with you. It's not worth testing that. I will remove it. :-)